sphinx/domains/c.py

```diff
@@ -3128,8 +3149,26 @@ def handle_signature(self, sig: str, signode: TextElement) -> ASTDeclaration:
 
         parser = DefinitionParser(sig, location=signode, config=self.env.config)
         try:
-            ast = self.parse_definition(parser)
-            parser.assert_end()
+            try:
+                ast = self.parse_definition(parser)
+                parser.assert_end()
+            except DefinitionError as eOrig:
+                if not self.env.config['c_allow_pre_v3']:
+                    raise
+                if self.objtype != 'type':
+                    raise
+                try:
+                    ast = self.parse_pre_v3_type_definition(parser)
+                    parser.assert_end()
+                except DefinitionError:
+                    raise eOrig
+                self.object_type = ast.objectType  # type: ignore
+                msg = "{}: Pre-v3 C type directive '.. c:type:: {}' converted to " \
+                      "'.. c:{}:: {}'." \
+                      "\nThe original parsing error was:\n{}"
+                msg = msg.format(RemovedInNextVersionWarning.__name__,
```

Comment:
Is it really needed to remove this code path? Would it be possible to wait before deprecating it, to give time to projects to first upgrade to Sphinx 3? With this PR, it's not possible to upgrade to Sphinx 3 and treat warnings as errors, since c_allow_pre_v3=True emits tons of warnings.

Reply:
With the added ``c_warn_on_allowed_pre_v3`` option you can now disable the warning.